Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertRebar.cs

```diff
@@ -1,175 +1,332 @@
+      {
+        appObj.Update(status: ApplicationObject.State.Failed, logItem: disciplineException.Message);
+      }
+      finally
+      {
+        if (rebar != null)
+        {
+          SetInstanceParameters(rebar, speckleRebar);
 
-      if (speckleRevitRebar != null)
-        SetInstanceParameters(rebar, speckleRevitRebar);
+          // set additional params
+          if (speckleRebar.barPositions > 0)
+          {
+            rebar.NumberOfBarPositions = speckleRebar.barPositions;
+          }
 
-      appObj.Update(status: ApplicationObject.State.Created, createdId: rebar.UniqueId, convertedItem: rebar);
+          // deleting instead of updating for now!
+          if (docObj != null)
+            Doc.Delete(docObj.Id);
+
+          appObj.Update(status: ApplicationObject.State.Created, createdId: rebar.UniqueId, convertedItem: rebar);
+        }
+      }
       return appObj;
     }
 
-    // need to test to see possible types of curves for rebar
-    private void ProcessRebarCurves(Rebar rebar, out List<DB.Curve> open, out List<DB.CurveArray> closed)
+    private RevitRebarGroup RebarToSpeckle(DB.Structure.Rebar revitRebar)
     {
-      var _open = new List<DB.Curve>();
-      var _closed = new List<DB.CurveArray>();
-
-      foreach (var curve in rebar.curves)
+      // skip freeform rebar for now: not supported by RevitRebarGroup class
+      // this is because freeform rebar with bent workshop has multiple shapes
+      if (revitRebar.GetAllRebarShapeIds().Count > 1)
       {
-        var array = CurveToNative(curve);
-        switch (curve)
-        {
-          case Polyline o:
-            if (o.closed)
-              _closed.Add(array);
-            else
-            {
-              var iterator = array.ForwardIterator();
-              while (iterator.MoveNext())
-                _open.Add(iterator.Current as DB.Curve);
-            }
-            break;
-          case Polycurve o:
-            if (o.closed)
-              _closed.Add(array);
-            else
-            {
-              var iterator = array.ForwardIterator();
-              while (iterator.MoveNext())
-                _open.Add(iterator.Current as DB.Curve);
-            }
-            break;
-          default:
-            _open.Add(array.get_Item(0));
-            break;
-        }
+        return null;
       }
 
-      open = _open; closed = _closed;
-    }
+      // get type
+      var type = revitRebar.Document.GetElement(revitRebar.GetTypeId()) as ElementType;
 
-    private Base RebarToSpeckle(DB.Structure.Rebar revitRebar)
-    {
-      // get rebar centerline curves using transform
-      var bars = revitRebar.GetCenterlineCurves(true, true, true, MultiplanarOption.IncludeOnlyPlanarCurves, revitRebar.NumberOfBarPositions - 1);
-      var curves = new List<ICurve>();
+      // get the rebar shape
+      var revitShape = revitRebar.Document.GetElement(revitRebar.GetShapeId()) as DB.Structure.RebarShape;
```

Comment:
Document is always valid? Can revitRebar be null? Does RebarShapeToSpeckle() accept null if it did not cast for some reason?

Reply:
Hmm good point, I'll add more robust conditionals here, even though those params should always be valid